contracts/LightClientBridge.sol

```diff
@@ -0,0 +1,282 @@
+        require(validatorPublicKeyMerkleProofs.length == requiredNumOfSignatures,
+            "Error: Number of validator public keys does not match required");
+
+        /**
+         * @dev Generate an array of numbers
+         */
+        uint256[] memory randomBitfield = Bitfield.randomNBitsFromPrior(getSeed(data), data.validatorClaimsBitfield,
+            requiredNumOfSignatures);
+
+        /**
+         *  @dev For each randomSignature, do:
+         */
+        for (uint256 i = 0; i < requiredNumOfSignatures; i++) {
+            /**
+             * @dev Check if validator in randomBitfield
+             */
+            require(
+                randomBitfield.isSet(validatorPositions[i]), "Error: Validator must be once in bitfield"
+            );
+
+            /**
+             * @dev Remove validator from randomBitfield such that no validator can appear twice in signatures
+             */
+            randomBitfield.clear(validatorPositions[i]);
+
+            /**
+             * @dev Check if merkle proof is valid
+             */
+            require(
+                validatorRegistry.checkValidatorInSet(validatorPublicKeys[i], validatorPositions[i],
+                validatorPublicKeyMerkleProofs[i]), "Error: Validator must be in validator set at correct position"
+            );
+
+            /**
+             * @dev Check if signature is correct
+             */
+            require(ECDSA.recover(payload, signatures[i]) == validatorPublicKeys[i], "Error: Invalid Signature");
+        }
+
+        /**
+         * @todo Release the sender stake as collateral
+         */
+        // TODO
+
+        /**
+         * @follow-up Do we need a try-catch block here?
+         */
+        processPayload(data.payload);
+
+        emit FinalVerificationSuccessful(msg.sender, payload, id);
+
+        /**
+         * @dev We no longer need the data held in state, so delete it for a gas refund
+         */
+        delete validationData[id];
+    }
+
+    /* Private Functions */
+
+    /**
+     * @notice Deterministically generates an array of numbers using the blockhash as a seed
+     * @dev Note that `blockhash(blockNum)` will only work for the 256 most recent blocks
+     * @dev Each generated number must be less than MAXIMUM_NUM_SIGNERS
+     * @param data a storage reference to the validationData struct
+     * @return onChainRandNums an array storing the random numbers generated inside this function
+     */
+    function getSeed(ValidationData storage data)
+        private
+        view
+        returns (uint256)
+    {
+        // @note Get payload.blocknumber, add BLOCK_WAIT_PERIOD
+        uint256 randomSeedBlockNum = data.blockNumber.add(BLOCK_WAIT_PERIOD);
+        // @note Create a hash seed from the block number
+        bytes32 randomSeedBlockHash = blockhash(randomSeedBlockNum);
+        //TODO: What happens if randomSeedBlockNum is too far in the past? Will we get an error/revert?
```

Comment:
yep then it will error. someone will need to create a new newSignatureCommitment again